### PR TITLE
[CPU][Zen] Route Int8 MoE inference through zentorch on AMD

### PR DESCRIPTION
**Pull request by ganeshr10**
## Purpose
Routes CPU Int8 W8A8 MoE on AMD Zen through zentorch: the expert GEMMs and activation run as a single torch.ops.zentorch.zentorch_fused_moe call, mirroring the existing zentorch W8A8 dense-linear integration. When zentorch's fused-MoE op is unavailable or the deployment config isn't one it supports selection falls through to the existing classes and behaviour elsewhere is unchanged.

- ZenCPUExpertsInt8 (cpu_moe.py): monolithic int8 experts class, channelwise weights with dynamic per-token activations, routing through the shared select_experts.

- Int8 MoE oracle: register ZenCPUExpertsInt8 in the CPU backend.

## Test Plan
gsm8k 5-shot on gpt-oss-20b W8A8 (compressed-tensors, channelwise weights with dynamic per-token activations), on an AMD Zen CPU with zentorch installed:

```lm_eval --model vllm --model_args pretrained=<gpt-oss-20b-BF16-w8a8>,dtype=bfloat16,max_model_len=4096 --tasks gsm8k --batch_size auto --trust_remote_code --num_fewshot 5 --apply_chat_template --log_samples --gen_kwargs "max_gen_toks=2048"```

The re-quantized checkpoint is in the process of being uploaded;
## Test Result
Verified GPT-OSS / Qwen3 MoE W8A8 inference on AMD Zen CPU. gsm8k 5-shot over all 1319 samples: flexible-extract exact_match 0.887 ± 0.0087. 

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

File: vllm/_custom_ops.py [hunk 1]
```diff
@@ -2499,6 +2539,17 @@ def topk_softmax(
...
+        _native_topk_score_then_select(
+            torch.softmax,
+            topk_weights,
+            topk_ids,
+            token_expert_indices,
+            gating_output,
+            renormalize,
+            e_score_correction_bias,
+        )
+        return
```
Comment (tlrmchlsmth) [inline vllm/_custom_ops.py]: This isn't the right way to integrate the `topk_softmax` implementation - you should be able to register a CPU dispatch key for the _moe_C ops so the existing router just works

File: vllm/model_executor/models/gpt_oss.py
Comment (tlrmchlsmth) [inline vllm/model_executor/models/gpt_oss.py:1]: Why are the changes to the weight loading needed? Do you have a particular model stub that these changes are needed for?

File: vllm/model_executor/models/gpt_oss.py [hunk 2]
```diff
@@ -263,6 +263,106 @@ def forward(
...
+    ep_rank_start: int,
+    ep_rank_end: int,
+    tp_rank: int,
+    tp_rank_start: int,
+    tp_rank_end: int,
+    per_rank_intermediate_size: int,
+) -> bool:
+    """Route LLM-Compressor per-expert keys (experts.experts.N.{gate,up,down}_proj.*)
+    into the stacked w13_*/w2_* params, honoring EP/TP. Returns True iff the
+    name matched a per-expert pattern (caller should `continue`)."""
```
Comment (tlrmchlsmth) [inline vllm/model_executor/models/gpt_oss.py]: should read `compressed-tensors` rather than `llm-compressor`, no?

**Review by tlrmchlsmth — CHANGES_REQUESTED** [review]

File: vllm/model_executor/models/gpt_oss.py (hunk 2 above)
Comment (ganeshr10) [reply vllm/model_executor/models/gpt_oss.py]: Fixed it

File: vllm/_custom_ops.py (hunk 1 above)
Comment (ganeshr10) [reply vllm/_custom_ops.py]: Fixed it.

File: vllm/model_executor/models/gpt_oss.py
Comment (ganeshr10) [reply vllm/model_executor/models/gpt_oss.py:1]: Yes for amd/gpt-oss-20b-BF16-w8a8-llmcompressor-v0.10.0.2 . Its checkpoint stores experts per-expert and per-projection

That's a different shape from the two layouts already handled OpenAI mxfp4 (fused gate_up_proj, _load_weights_mxfp4) and AMD-quark (per-expert with gate+up pre-fused into w13, _load_weights_quark). compressed-tensors keeps gate/up/down as separate per-expert Linears (llm-compressor expands the fused gpt-oss experts before quantizing each), so _load_weights_other had no path for them.

**Comment by amukho** [comment]
hi @xuechendi this PR is adding support for routing the w8a8 fused MoE op through the custom kernel for AMD CPUs. It would be great if you could review and help merge this change
cc @tlrmchlsmth

**Comment by amukho** [comment]
hi @dllehr-amd this PR is adding support for routing the w8a8 fused MoE op through the custom kernel for AMD CPUs. It would be great if you could review and help merge this change

**Comment by aadwived** [comment]
Hello

@tlrmchlsmth : Can we get a review on the updated PR? We have responded to the points raised earlier.
@xuechendi : Can we get a review for this PR?

CC : @amukho

**Comment by Fangzhou-Ai** [comment]
Is zentorch fused MoE faster than the current cpu fused moe?

**Comment by ganeshr10** [comment]
Hi @Fangzhou-Ai , 

- Mixtral-8x7B: zentorch is ~18-23% faster than the generic CPU fused MoE 
- Qwen3-30B / Qwen3.6-35B: zentorch is ~8-10% slower than the generic kernel.
- gpt-oss-20b: the generic CPU fused MoE doesn't support it (SwiGLU-OAI + expert bias) so I couldn't get data without zentorch.

image

File: vllm/model_executor/models/gpt_oss.py
Comment (ganeshr10) [reply vllm/model_executor/models/gpt_oss.py:1]: The weight loading part has been moved to a separate utils file.

File: tests/models/quantization/test_gpt_oss_per_expert_loader.py [hunk 3]
```diff
@@ -0,0 +1,233 @@
+# SPDX-License-Identifier: Apache-2.0
+# SPDX-FileCopyrightText: Copyright contributors to the vLLM project
+"""Pure-CPU smoke tests for _load_per_expert_moe_weight in gpt_oss.py:
+gate/up w13 packing, w2 TP-slicing, the w2_bias rank-0 rule, and
+off-rank (EP) / non-MoE / missing-param handling."""
+
+import pytest
+import torch
+
+from vllm.model_executor.models.gpt_oss import _load_per_expert_moe_weight
```
Comment (dllehr-amd) [inline tests/models/quantization/test_gpt_oss_per_expert_loader.py]: I believe this has moved since a refactor.  Can you address this?  I think it's under `moe_loading_utils.py`

Comment (ganeshr10) [reply tests/models/quantization/test_gpt_oss_per_expert_loader.py]: Updated the path.

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe/compressed_tensors_moe_w8a8_int8.py [hunk 4]
```diff
@@ -141,6 +164,51 @@ def create_weights(
...
+        # _load_w13 produces a half-split layout. Reorder w13 to interleaved
+        # once, gated on Zen CPU so other backends keep the half-split layout.
+        # This must happen before convert_to_int8_moe_kernel_format below.
+        _act = getattr(layer, "activation", None)
+        _act_str = getattr(_act, "value", _act)
+        if (
+            current_platform.is_zen_cpu()
+            and isinstance(_act_str, str)
+            and _act_str.lower() == "swigluoai"
+        ):
```
Comment (tlrmchlsmth) [inline vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe/compressed_tensors_moe_w8a8_int8.py:178]: It looks like we will incorrectly permute the weights when we're running on Zen CPUs but the zentorch backend is not available. Could you take a look?

File: vllm/model_executor/layers/quantization/compressed_tensors/moe_loading_utils.py
Comment (tlrmchlsmth) [inline vllm/model_executor/layers/quantization/compressed_tensors/moe_loading_utils.py:1]: This appears to be a model-specific compatibility shim. Ideally we don't need this.

The layers in `amd/gpt-oss-20b-BF16-w8a8-llmcompressor-v0.10.0.2` have names like `model.layers.0.mlp.experts.experts.2.down_proj.weight`. I think it would be better to fix the checkpoint instead

**Review by tlrmchlsmth — COMMENTED** [review]
It looks like the https://huggingface.co/amd/gpt-oss-20b-BF16-w8a8-llmcompressor-v0.10.0.2 checkpoint is a little wacky, so I'm wondering if the weight loading changes rather than something needed for the `compressed-tensors` format

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe/compressed_tensors_moe_w8a8_int8.py (hunk 4 above)
Comment (ganeshr10) [reply vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe/compressed_tensors_moe_w8a8_int8.py:178]: Done. The gate now uses the selected experts class (self.experts_cls is ZenCPUExpertsInt8)

File: vllm/model_executor/layers/quantization/compressed_tensors/moe_loading_utils.py
Comment (ganeshr10) [reply vllm/model_executor/layers/quantization/compressed_tensors/moe_loading_utils.py:1]: Removed it

**Comment by ganeshr10** [comment]
> It looks like the https://huggingface.co/amd/gpt-oss-20b-BF16-w8a8-llmcompressor-v0.10.0.2 checkpoint is a little wacky, so I'm wondering if the weight loading changes rather than something needed for the `compressed-tensors` format

Yes, those were specific to that checkpoint, not to compressed-tensors. The checkpoint was quantized using llm-compressor 0.10, which added the extra experts.experts.N nesting and left gate/up/down split per expert.
Re-quantized with llm-compressor 0.13 and hence the moe_loading_utils.py is no longer needed

**Review by amukho — APPROVED** [review]
Looks good to me...

**Review by tlrmchlsmth — COMMENTED** [review]
Looks good but one last thing - could you please ensure that there is a test exercising `ZenCPUExpertsInt8` in CI?

**Comment by ganeshr10** [comment]
> Looks good but one last thing - could you please ensure that there is a test exercising `ZenCPUExpertsInt8` in CI?

Done. I've added a test.

**Review by tlrmchlsmth — APPROVED** [review]

**Comment by ganeshr10** [comment]
@tlrmchlsmth Could you please add the `ready` label so full CI runs?
